### PR TITLE
Test the CSV read with column types specified

### LINKED ISSUE #23133: [Dataset] Unable to read scientific notation such as '1E+05' in CSV.
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Others

### Issue Severity

Medium: It contributes to significant difficulty to complete my task but I work arounds and get it resolved.

### What happened + What you expected to happen

```
import pandas as pd
from pyarrow.csv import read_csv
from ray import data


pd.read_csv("/mnt/d/Projects/fusion-platform/example/creditcard.csv")

pa = read_csv("/mnt/d/Projects/fusion-platform/example/creditcard.csv")

ds = data.read_csv("/mnt/d/Projects/fusion-platform/example/creditcard.csv")
```

Ray Dataset seems failed to read from the famous credit card fraud detection dataset. https://www.kaggle.com/mlg-ulb/creditcardfraud

While for Pandas and PyArrow (of which Ray dataset relies), the dataset can be load without any problem. 

### Versions / Dependencies

1.10.0

### Reproduction script

Download CSV from https://www.kaggle.com/mlg-ulb/creditcardfraud

Run 
```
import pandas as pd
from pyarrow.csv import read_csv
from ray import data


pd.read_csv("/mnt/d/Projects/fusion-platform/example/creditcard.csv")

pa = read_csv("/mnt/d/Projects/fusion-platform/example/creditcard.csv")

ds = data.read_csv("/mnt/d/Projects/fusion-platform/example/creditcard.csv")
```

Ray data set part will complain

```
(remote_read pid=25419) pyarrow.lib.ArrowInvalid: In CSV column #0: Row #153760: CSV conversion error to int64: invalid value '1e+05'
```

### Anything else

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### LINKED ISSUE #23448: [Datasets] [Bug]  AIR dogfooding- Reading csv with ray.data fails, but works with pandas
[UCI_Credit_Card.csv](https://github.com/ray-project/ray/files/8338068/UCI_Credit_Card.csv)

Reading the attached csv file with `ray.data` fails with the following error

```
/Users/kai/.pyenv/versions/3.7.7/bin/python /Users/kai/coding/ray/python/ray/ml/examples/seldon_example_cc.py
2022-03-23 16:18:14,529	INFO services.py:1462 -- View the Ray dashboard at[ http://127.0.0.1:8265](http://127.0.0.1:8265/)
(remote_read pid=36776) 2022-03-23 16:18:18,972	INFO worker.py:449 -- Task failed with retryable exception: TaskID(32d950ec0ccf9d2affffffffffffffffffffffff01000000).
(remote_read pid=36776) Traceback (most recent call last):
(remote_read pid=36776)   File "python/ray/_raylet.pyx", line 663, in ray._raylet.execute_task
(remote_read pid=36776)   File "python/ray/_raylet.pyx", line 667, in ray._raylet.execute_task
(remote_read pid=36776)   File "/Users/kai/coding/ray/python/ray/data/read_api.py", line 250, in remote_read
(remote_read pid=36776)     block = task()
(remote_read pid=36776)   File "/Users/kai/coding/ray/python/ray/data/datasource/datasource.py", line 155, in __call__
(remote_read pid=36776)     for block in result:
(remote_read pid=36776)   File "/Users/kai/coding/ray/python/ray/data/datasource/file_based_datasource.py", line 316, in read_files
(remote_read pid=36776)     for data in read_stream(f, read_path, **reader_args):
(remote_read pid=36776)   File "/Users/kai/coding/ray/python/ray/data/datasource/csv_datasource.py", line 35, in _read_stream
(remote_read pid=36776)     batch = reader.read_next_batch()
(remote_read pid=36776)   File "pyarrow/ipc.pxi", line 543, in pyarrow.lib.RecordBatchReader.read_next_batch
(remote_read pid=36776)   File "pyarrow/error.pxi", line 97, in pyarrow.lib.check_status
(remote_read pid=36776) pyarrow.lib.ArrowInvalid: In CSV column #12: CSV conversion error to int64: invalid value '1e+05'
(remote_read pid=36776) 2022-03-23 16:18:19,130	INFO worker.py:449 -- Task failed with retryable exception: TaskID(32d950ec0ccf9d2affffffffffffffffffffffff01000000).
(remote_read pid=36776) Traceback (most recent call last):
(remote_read pid=36776)   File "python/ray/_raylet.pyx", line 663, in ray._raylet.execute_task
(remote_read pid=36776)   File "python/ray/_raylet.pyx", line 667, in ray._raylet.execute_task
(remote_read pid=36776)   File "/Users/kai/coding/ray/python/ray/data/read_api.py", line 250, in remote_read
(remote_read pid=36776)     block = task()
(remote_read pid=36776)   File "/Users/kai/coding/ray/python/ray/data/datasource/datasource.py", line 155, in __call__
(remote_read pid=36776)     for block in result:
(remote_read pid=36776)   File "/Users/kai/coding/ray/python/ray/data/datasource/file_based_datasource.py", line 316, in read_files
(remote_read pid=36776)     for data in read_stream(f, read_path, **reader_args):
(remote_read pid=36776)   File "/Users/kai/coding/ray/python/ray/data/datasource/csv_datasource.py", line 35, in _read_stream
(remote_read pid=36776)     batch = reader.read_next_batch()
(remote_read pid=36776)   File "pyarrow/ipc.pxi", line 543, in pyarrow.lib.RecordBatchReader.read_next_batch
(remote_read pid=36776)   File "pyarrow/error.pxi", line 97, in pyarrow.lib.check_status
(remote_read pid=36776) pyarrow.lib.ArrowInvalid: In CSV column #12: CSV conversion error to int64: invalid value '1e+05'
(pid=36765) 
(remote_read pid=36776) 2022-03-23 16:18:20,144	INFO worker.py:449 -- Task failed with retryable exception: TaskID(32d950ec0ccf9d2affffffffffffffffffffffff01000000).
(remote_read pid=36776) Traceback (most recent call last):
(remote_read pid=36776)   File "python/ray/_raylet.pyx", line 663, in ray._raylet.execute_task
(remote_read pid=36776)   File "python/ray/_raylet.pyx", line 667, in ray._raylet.execute_task
(remote_read pid=36776)   File "/Users/kai/coding/ray/python/ray/data/read_api.py", line 250, in remote_read
(remote_read pid=36776)     block = task()
(remote_read pid=36776)   File "/Users/kai/coding/ray/python/ray/data/datasource/datasource.py", line 155, in __call__
(remote_read pid=36776)     for block in result:
(remote_read pid=36776)   File "/Users/kai/coding/ray/python/ray/data/datasource/file_based_datasource.py", line 316, in read_files
(remote_read pid=36776)     for data in read_stream(f, read_path, **reader_args):
(remote_read pid=36776)   File "/Users/kai/coding/ray/python/ray/data/datasource/csv_datasource.py", line 35, in _read_stream
(remote_read pid=36776)     batch = reader.read_next_batch()
(remote_read pid=36776)   File "pyarrow/ipc.pxi", line 543, in pyarrow.lib.RecordBatchReader.read_next_batch
(remote_read pid=36776)   File "pyarrow/error.pxi", line 97, in pyarrow.lib.check_status
(remote_read pid=36776) pyarrow.lib.ArrowInvalid: In CSV column #12: CSV conversion error to int64: invalid value '1e+05'
Traceback (most recent call last):
  File "/Users/kai/coding/ray/python/ray/ml/examples/seldon_example_cc.py", line 9, in <module>
    dataset = load_data()
  File "/Users/kai/coding/ray/python/ray/ml/examples/seldon_example_cc.py", line 5, in load_data
    return ray.data.read_csv(path)
  File "/Users/kai/coding/ray/python/ray/data/read_api.py", line 481, in read_csv
    **arrow_csv_args,
  File "/Users/kai/coding/ray/python/ray/data/read_api.py", line 286, in read_datasource
    block_list.ensure_schema_for_first_block()
  File "/Users/kai/coding/ray/python/ray/data/impl/block_list.py", line 197, in ensure_schema_for_first_block
    schema = ray.get(get_schema.remote(block))
  File "/Users/kai/coding/ray/python/ray/_private/client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/Users/kai/coding/ray/python/ray/worker.py", line 1809, in get
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(ArrowInvalid): ray::_get_schema() (pid=36776, ip=127.0.0.1)
  At least one of the input arguments for this task could not be computed:
ray.exceptions.RayTaskError: ray::remote_read() (pid=36776, ip=127.0.0.1)
  File "/Users/kai/coding/ray/python/ray/data/read_api.py", line 250, in remote_read
    block = task()
  File "/Users/kai/coding/ray/python/ray/data/datasource/datasource.py", line 155, in __call__
    for block in result:
  File "/Users/kai/coding/ray/python/ray/data/datasource/file_based_datasource.py", line 316, in read_files
    for data in read_stream(f, read_path, **reader_args):
  File "/Users/kai/coding/ray/python/ray/data/datasource/csv_datasource.py", line 35, in _read_stream
    batch = reader.read_next_batch()
  File "pyarrow/ipc.pxi", line 543, in pyarrow.lib.RecordBatchReader.read_next_batch
  File "pyarrow/error.pxi", line 97, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: In CSV column #12: CSV conversion error to int64: invalid value '1e+05'
(remote_read pid=36776) 2022-03-23 16:18:21,126	INFO worker.py:449 -- Task failed with retryable exception: TaskID(32d950ec0ccf9d2affffffffffffffffffffffff01000000).
(remote_read pid=36776) Traceback (most recent call last):
(remote_read pid=36776)   File "python/ray/_raylet.pyx", line 663, in ray._raylet.execute_task
(remote_read pid=36776)   File "python/ray/_raylet.pyx", line 667, in ray._raylet.execute_task
(remote_read pid=36776)   File "/Users/kai/coding/ray/python/ray/data/read_api.py", line 250, in remote_read
(remote_read pid=36776)     block = task()
(remote_read pid=36776)   File "/Users/kai/coding/ray/python/ray/data/datasource/datasource.py", line 155, in __call__
(remote_read pid=36776)     for block in result:
(remote_read pid=36776)   File "/Users/kai/coding/ray/python/ray/data/datasource/file_based_datasource.py", line 316, in read_files
(remote_read pid=36776)     for data in read_stream(f, read_path, **reader_args):
(remote_read pid=36776)   File "/Users/kai/coding/ray/python/ray/data/datasource/csv_datasource.py", line 35, in _read_stream
(remote_read pid=36776)     batch = reader.read_next_batch()
(remote_read pid=36776)   File "pyarrow/ipc.pxi", line 543, in pyarrow.lib.RecordBatchReader.read_next_batch
(remote_read pid=36776)   File "pyarrow/error.pxi", line 97, in pyarrow.lib.check_status
(remote_read pid=36776) pyarrow.lib.ArrowInvalid: In CSV column #12: CSV conversion error to int64: invalid value '1e+05'
```

Pandas is able to successfully read the csv file.

There are workarounds to this so it is not fully blocking (such as using Dask on Ray and then converting to Ray Datasets), but it is difficult to overcome and persistent with this particular dataset.

### PR DESCRIPTION
**Pull request by jianoaix**
## Why are these changes needed?

Make sure users can read csv with columns types specified.
Users may want to do this because sometimes PyArrow's type inference doesn't work as intended, in which case users can step in and work around the type inference.

## Related issue number

Closes #23133
Closes #23448

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ y] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by clarkzinzow — APPROVED** [review]
Nice test!

File: python/ray/data/tests/test_dataset_formats.py [hunk 1]
```diff
@@ -2403,6 +2403,33 @@ def test_csv_read_no_header(shutdown_only, tmp_path):
...
+
+    # Parsing scientific notation in double shoud work.
+    ds = ray.data.read_csv(
+        file_path,
+        convert_options=csv.ConvertOptions(
+            column_types={"one": "float64", "two": "string"}),
+    )
+    expected_df = pd.DataFrame(
+            {"one": [1.0, 2.0, 3e+5], "two": ["a", "b", "c"]})
+    assert df.equals(expected_df)
```
Comment (clarkzinzow) [inline python/ray/data/tests/test_dataset_formats.py:2430]:
```suggestion
    assert ds.to_pandas().equals(expected_df)
```

Comment (jianoaix) [reply python/ray/data/tests/test_dataset_formats.py:2430]: Done, thanks!

**Comment by scv119** [comment]
> python/ray/data/tests/test_dataset_formats.py:2424:5: F841 local variable 'ds' is assigned to but never used

**Review by scv119 — APPROVED** [review]

**Comment by jianoaix** [comment]
Lint fixed, can you merge?

**Pull request merged**